src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,176 @@
 package core.basesyntax;
 
+import core.basesyntax.service.ObjectService;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(tail, value, null);
+        if (head == null) {
+            head = node;
+        }
+        if (tail != null) {
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if ((index == 0 && size == 0) || size == index) {
+            add(value);
+            return;
+        }
+        checkIndexOfBoundsException(index);
+        Node<T> foundedNode = getNodeByIndex(index);
+        if (foundedNode != null) {
+            insertNewNode(value, foundedNode);
+        } else {
+            createNewNodeAtTail(value);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOfBoundsException(index);
+        Node<T> foundedNode = getNodeByIndex(index);
+        return foundedNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexOfBoundsException(index);
+        Node<T> foundedNode = getNodeByIndex(index);
+        T oldItem = foundedNode.item;
+        foundedNode.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOfBoundsException(index);
+        Node<T> foundedNode = getNodeByIndex(index);
+        unlink(foundedNode);
+        return foundedNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            return false;
+        }
+        Node<T> foundedNode = getNodeByValue(object);
+        if (foundedNode != null) {
+            unlink(foundedNode);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexOfBoundsException(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " doesn't exist");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < size / 2 || index < 2) {
```

Comment:
its enough to check only devision on half
```suggestion
        if (index < size / 2) {
```